File: 3. Behavioral Design Patterns/Observer/Typescript/ObserverExample.ts
// Observer Pattern Example in TypeScript
// WeatherStation notifies Display observers when temperature changes

interface Observer { update(temperature: number): void }
interface Subject { attach(o: Observer): void; detach(o: Observer): void; notify(): void }

class WeatherStation implements Subject {
  private observers: Observer[] = []
  private temperature = 0
  attach(o: Observer){ this.observers.push(o) }
  detach(o: Observer){ this.observers = this.observers.filter(obs => obs !== o) }
  notify(){ for(const o of this.observers) o.update(this.temperature) }
  setTemperature(temp: number){ this.temperature = temp; this.notify() }
}

class PhoneDisplay implements Observer { update(temp: number){ console.log(`Phone: Temperature is ${temp}°C`) } }
class WindowDisplay implements Observer { update(temp: number){ console.log(`Window: Temperature is ${temp}°C`) } }

;(function main(){
  const station = new WeatherStation()
  const phone = new PhoneDisplay()
  const window = new WindowDisplay()
  station.attach(phone)
  station.attach(window)
  station.setTemperature(25.5)
  station.setTemperature(30.2)
})()

export {}

File: 3. Behavioral Design Patterns/Observer/Java/ObserverExample.java
// Observer Pattern Example in Java
// WeatherStation notifies Display observers when temperature changes

import java.util.*;

interface Observer { void update(float temperature); }
interface Subject { void attach(Observer o); void detach(Observer o); void notifyObservers(); }

class WeatherStation implements Subject {
    private final List<Observer> observers = new ArrayList<>();
    private float temperature;
    public void attach(Observer o){ observers.add(o); }
    public void detach(Observer o){ observers.remove(o); }
    public void notifyObservers(){ for (Observer o : observers) o.update(temperature); }
    public void setTemperature(float temp){ temperature = temp; notifyObservers(); }
}

class PhoneDisplay implements Observer { public void update(float temp){ System.out.printf("Phone: Temperature is %.1f°C%n", temp); } }
class WindowDisplay implements Observer { public void update(float temp){ System.out.printf("Window: Temperature is %.1f°C%n", temp); } }

public class ObserverExample {
    public static void main(String[] args){
        WeatherStation station = new WeatherStation();
        PhoneDisplay phone = new PhoneDisplay();
        WindowDisplay window = new WindowDisplay();
        station.attach(phone);
        station.attach(window);
        station.setTemperature(25.5f);
        station.setTemperature(30.2f);
    }
}
